File: ts/src/graph.ts
/* Copyright (c) 2025 Richard Rodger, MIT License */


import type { Val } from './type'

import { cmpCodePoint } from './keyorder'


export type Edge = {
  from: string
  key: string
  to: string
  // Where the link is, as a `$.dotted.path`, so a report can point at
  // it.
  at: string
  hidden?: true
}

export type Graph = {
  edges: Edge[]
  disjunct?: string[]
}


const formatPath = (path: string[]): string =>
  0 === path.length ? '$' : '$.' + path.join('.')


// Digits-only segments are list indices, which is exactly how the rest
// of the engine spells them.
const isIndex = (seg: string): boolean => /^[0-9]+$/.test(seg)


const cut = (
  at: string[], relkey: string | undefined
): { from: string, key: string } => {
  if (undefined !== relkey) {
    for (let i = at.length - 1; 0 <= i; i--) {
      if (at[i] === relkey) {
        return { from: formatPath(at.slice(0, i)), key: relkey }
      }
    }
  }
  let i = at.length - 1
  for (; 0 <= i && isIndex(at[i]); i--) { }
  return 0 > i
    ? { from: formatPath([]), key: relkey ?? '' }
    : { from: formatPath(at.slice(0, i)), key: relkey ?? at[i] }
}


export function graphOf(root: Val): Graph {
  const edges: Edge[] = []
  const disjunct: string[] = []

  // ONE WALK, with `undecided` saying which side of ADR-007 it is on:
  // below an unresolved disjunction every link is a position the
  // document has not decided, and nothing there is an edge.
  const visit = (
    node: any, path: string[], ancestors: Set<any>, hidden: boolean,
    undecided: boolean
  ): void => {
    if (null == node || true !== node.isVal || ancestors.has(node)) {
      return
    }

    hidden = hidden || true === node.mark?.hide

    const link = node.link
    if (null != link) {
      if (undecided) {
        disjunct.push(formatPath(path))
      }
      else {
        const { from, key } = cut(path, node.relkey as string | undefined)
        const edge: Edge = { from, key, to: link, at: formatPath(path) }
        if (hidden) {
          edge.hidden = true
        }
        edges.push(edge)
      }
    }

    // A graph atom is TRANSPARENT here (RELATIONS P2): it carries the
    // field's value at the field's own position, and the graph is about
    // the value.
    if (true === node.isGraphAtom && undefined !== node.held) {
      visit(node.held, path, ancestors, hidden, undecided)
    }

    if (true === node.isConjunct && Array.isArray(node.peg)) {
      ancestors.add(node)
      for (const term of node.peg) {
        visit(term, path, ancestors, hidden, undecided)
      }
      ancestors.delete(node)
    }

    if (true === node.isDisjunct && Array.isArray(node.peg)) {
      ancestors.add(node)
      for (const arm of node.peg) {
        visit(arm, path, ancestors, hidden, true)
      }
      ancestors.delete(node)
    }

    if ((true === node.isMap || true === node.isList) && null != node.peg) {
      ancestors.add(node)
      for (const k of Object.keys(node.peg)) {
        visit(node.peg[k], [...path, k], ancestors, hidden, undecided)
      }
      ancestors.delete(node)
    }
  }

  visit(root, [], new Set(), false, false)

  edges.sort((a, b) => cmpCodePoint(a.at, b.at))

  if (0 < disjunct.length) {
    return { edges, disjunct: [...new Set(disjunct)].sort(cmpCodePoint) }
  }
  return { edges }
}

File: go/graph.go
/* Copyright (c) 2025 Richard Rodger, MIT License */

package aontu

import (
	"sort"
	"strings"
)


// Edge is one checked link. Mirrors Edge in ts/src/graph.ts.
type Edge struct {
	From string `json:"from"`
	Key string `json:"key"`
	To string `json:"to"`
	// At is where the link is, as a `$.dotted.path`.
	At string `json:"at"`
	Hidden bool `json:"hidden,omitempty"`
}

type Graph struct {
	Edges []Edge `json:"edges"`
	Disjunct []string `json:"disjunct,omitempty"`
}

func graphPath(path []string) string {
	if 0 == len(path) {
		return "$"
	}
	return "$." + strings.Join(path, ".")
}

func cutEdge(at []string, relkey string) (string, string) {
	if "" != relkey {
		for i := len(at) - 1; 0 <= i; i-- {
			if at[i] == relkey {
				return graphPath(at[:i]), relkey
			}
		}
	}
	i := len(at) - 1
	for ; 0 <= i && allDigits(at[i]); i-- {
	}
	if 0 > i {
		return graphPath(nil), relkey
	}
	key := relkey
	if "" == key {
		key = at[i]
	}
	return graphPath(at[:i]), key
}

func GraphOf(root Val) Graph {
	edges := []Edge{}
	disjunct := []string{}
	ancestors := map[Val]bool{}

	// ONE WALK, with `undecided` saying which side of ADR-007 it is on:
	// below an unresolved disjunction every link is a position the
	// document has not decided, and nothing there is an edge.
	var visit func(node Val, path []string, hidden, undecided bool)
	visit = func(node Val, path []string, hidden, undecided bool) {
		if nil == node || ancestors[node] {
			return
		}

		hidden = hidden || node.markedHide()

		if link := node.linkAddr(); "" != link && undecided {
			disjunct = append(disjunct, graphPath(path))
		} else if "" != link {
			relkey := ""
			if bb, ok := node.(interface{ relKey() string }); ok {
				relkey = bb.relKey()
			}
			from, key := cutEdge(path, relkey)
			edges = append(edges, Edge{
				From:   from,
				Key:    key,
				To:     link,
				At:     graphPath(path),
				Hidden: hidden,
			})
		}

		// A graph atom is TRANSPARENT here (RELATIONS P2): it carries
		// the field's value at the field's own position, and the graph
		// is about the value.
		if ga, ok := node.(*GraphAtomVal); ok && nil != ga.held {
			visit(ga.held, path, hidden, undecided)
		}

		switch n := node.(type) {
		// An unresolved conjunction holds its terms at the SAME
		// position; every link among them is written there, and is an
		// edge.
		case *ConjunctVal:
			ancestors[node] = true
			for _, t := range n.peg {
				visit(t, path, hidden, undecided)
			}
			delete(ancestors, node)
		case *DisjunctVal:
			ancestors[node] = true
			for _, arm := range n.peg {
				visit(arm, path, hidden, true)
			}
			delete(ancestors, node)
		case *MapVal:
			ancestors[node] = true
			for _, k := range n.keys {
				visit(n.peg[k], append(cp(path), k), hidden, undecided)
			}
			delete(ancestors, node)
		case *ListVal:
			ancestors[node] = true
			for i, e := range n.peg {
				visit(e, append(cp(path), itoa(i)), hidden, undecided)
			}
			delete(ancestors, node)
		}
	}

	visit(root, nil, false, false)

	sort.Slice(edges, func(i, j int) bool { return edges[i].At < edges[j].At })

	if 0 < len(disjunct) {
		sort.Strings(disjunct)
		unique := []string{}
		for i, p := range disjunct {
			if 0 == i || disjunct[i-1] != p {
				unique = append(unique, p)
			}
		}
		return Graph{Edges: edges, Disjunct: unique}
	}
	return Graph{Edges: edges}
}
